x-pack/platform/plugins/shared/observability_ai_assistant/server/service/anonymization/deanonymize_text.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { HASH_REGEX } from '../../../common/utils/anonymization/redaction';
+import { DetectedEntity, DetectedEntityType } from '../../../common/types';
+
+/**
+ * Replaces hash placeholders in text with their original values
+ *
+ * Takes text containing hash placeholders like "{hash123}" and replaces
+ * them with their original values from the provided hash map. Also generates
+ * entity metadata for each replaced value.
+ *
+ * @param contentWithHashes - Text containing hash placeholders
+ * @param hashMap - Map of hash values to their original entity information
+ * @returns Object containing deanonymized text and detected entities
+ */
+export function deanonymizeText(
+  contentWithHashes: string,
+  hashMap: Map<string, { value: string; class_name: string; type: DetectedEntityType }>
+) {
+  const detectedEntities: DetectedEntity[] = [];
+  let unhashedText = '';
+  let cursor = 0;
+
+  let match: RegExpExecArray | null;
+  while ((match = HASH_REGEX.exec(contentWithHashes)) !== null) {
```

Comment:
I think we should inline the regex here, as exec is stateful and I would like to prevent any confusion around `lastIndex` and what happens when part of this code fails and the regex is executed again after.